pkg/core/blockchain.go

```diff
@@ -677,36 +677,111 @@ func (bc *Blockchain) GetStateModule() blockchainer.StateRoot {
+			blockdone <- err
+			return
+		}
+		writeBuf.Reset()
 
+		if err := kvcache.StoreAsCurrentBlock(block, writeBuf); err != nil {
+			blockdone <- err
+			return
+		}
+		writeBuf.Reset()
+
+		for _, tx := range block.Transactions {
+			if err := kvcache.StoreAsTransaction(tx, block.Index, writeBuf); err != nil {
+				blockdone <- err
+			}
+
+			writeBuf.Reset()
+			if bc.config.P2PSigExtensions {
+				for _, attr := range tx.GetAttributes(transaction.ConflictsT) {
+					hash := attr.Value.(*transaction.Conflicts).Hash
+					dummyTx := transaction.NewTrimmedTX(hash)
+					dummyTx.Version = transaction.DummyVersion
+					if err := kvcache.StoreAsTransaction(dummyTx, block.Index, writeBuf); err != nil {
+						blockdone <- fmt.Errorf("failed to store conflicting transaction %s for transaction %s: %w", hash.StringLE(), tx.Hash().StringLE(), err)
+						return
+					}
+					writeBuf.Reset()
+				}
+			}
+		}
+		if bc.config.RemoveUntraceableBlocks {
+			if block.Index > bc.config.MaxTraceableBlocks {
+				index := block.Index - bc.config.MaxTraceableBlocks // is at least 1
+				err := kvcache.DeleteBlock(bc.headerHashes[index], writeBuf)
+				if err != nil {
+					bc.log.Warn("error while removing old block",
+						zap.Uint32("index", index),
+						zap.Error(err))
+				}
+				writeBuf.Reset()
+			}
+		}
+		_, err := kvcache.Persist()
+		if err != nil {
+			blockdone <- err
+		}
+		close(blockdone)
+	}()
+	go func() {
+		kvcache := dao.NewCached(cache)
+		var writeBuf = io.NewBufBinWriter()
+		var err error
+		var appendBlock bool
+		for {
+			aer, ok := <-aerchan
+			if !ok {
+				break
+			}
+			if aer.Container == block.Hash() && appendBlock {
+				err = kvcache.AppendAppExecResult(aer, writeBuf)
+			} else {
+				err = kvcache.PutAppExecResult(aer, writeBuf)
+				if aer.Container == block.Hash() {
+					appendBlock = true
+				}
+			}
+			if aer.Execution.VMState == vm.HaltState {
+				for j := range aer.Execution.Events {
+					bc.handleNotification(&aer.Execution.Events[j], kvcache, block, aer.Container)
+				}
+			}
+			if err != nil {
+				err = fmt.Errorf("failed to store tx exec result: %w", err)
+				break
+			}
+			writeBuf.Reset()
+		}
+		if err != nil {
+			aerdone <- err
+			return
+		}
+		_, err = kvcache.Persist()
+		if err != nil {
+			aerdone <- err
+		}
+		close(aerdone)
+	}()
 	aer, err := bc.runPersist(bc.contracts.GetPersistScript(), block, cache, trigger.OnPersist)
 	if err != nil {
 		return fmt.Errorf("onPersist failed: %w", err)
 	}
 	appExecResults = append(appExecResults, aer)
-	err = cache.PutAppExecResult(aer, writeBuf)
-	if err != nil {
-		return fmt.Errorf("failed to store onPersist exec result: %w", err)
-	}
-	writeBuf.Reset()
+	aerchan <- aer
 
 	for _, tx := range block.Transactions {
-		if err := cache.StoreAsTransaction(tx, block.Index, writeBuf); err != nil {
-			return err
-		}
-		writeBuf.Reset()
-
 		systemInterop := bc.newInteropContext(trigger.Application, cache, block, tx)
```

Comment:
Are you sure that the first step (saving block/transactions to the DB) is completely independent of the second step (executing blocks/transactions)? Consider the case of a block with several transactions A, B, C and the following situations:
1. Transaction B calls native Ledger's `getTransaction` where the argument is the hash of C. Sequential implementation of the `storeBlock` guarantees that `Null` will be returned from that method because C is stored exactly after B invocation. Concurrent implementation does not guarantee that because C may be stored (and persisted) before B invocation.
2. The similar thing may happen when transaction B calls native Ladger's `getTransaction` with A.Hash argument. Sequential implementation guarantees that proper A state will be returned from this call. Concurrent implementation does not.

So could you explain if I'm wrong?

Reply:
https://github.com/nspcc-dev/neo-go/blob/master/pkg/core/native/ledger.go#L146-L152

Reply:
Current (as of `storeBlock`) block is not traceable.